src/audio/codec_adapter/codec/waves.c

```diff
@@ -0,0 +1,726 @@
+	 * there migh be more than one struct codec_param inside cfg->data, glued back to back
+	 */
+	for (index = 0; index < cfg->size && (!ret); param_number++) {
+		uint32_t param_data_size;
+
+		param = (struct codec_param *)((char *)cfg->data + index);
+		param_data_size = param->size - sizeof(param->size) - sizeof(param->id);
+
+		comp_info(dev, "waves_codec_configure() param num %d id %d size %d",
+			param_number, param->id, param->size);
+
+		switch (param->id) {
+		case PARAM_NOP:
+			comp_info(dev, "waves_codec_configure() NOP");
+			ret = 0;
+			break;
+		case PARAM_MESSAGE:
+			ret = waves_effect_message(dev, param->data, param_data_size);
+			break;
+		case PARAM_REVISION:
+			ret = waves_effect_revision(dev);
+			break;
+		default:
+			ret = -EIO;
+			break;
+		}
+
+		index += param->size;
+	}
+
+	if (ret)
+		comp_err(dev, "waves_codec_configure() error %d", ret);
+	else
+		comp_dbg(dev, "waves_codec_configure() done");
+
+	return ret;
+}
+
+static int waves_effect_setup_config(struct comp_dev *dev)
+{
+	/* apply setup config if present */
+	struct codec_data *codec = comp_get_codec(dev);
+	int ret = 0;
+
+	comp_dbg(dev, "waves_effect_setup_config() start");
+
+	if (!codec->s_cfg.avail && !codec->s_cfg.size) {
+		comp_err(dev, "waves_effect_startup_config() no setup config");
+		return -EIO;
+	}
+
+	if (!codec->s_cfg.avail) {
+		comp_warn(dev, "waves_effect_startup_config() no new setup config, using old");
+		codec->s_cfg.avail = true;
+	}
+
+	ret = waves_effect_config(dev, CODEC_CFG_SETUP);
+	codec->s_cfg.avail = false;
+
+	comp_dbg(dev, "waves_effect_setup_config() done");
+	return ret;
+}
+
+int waves_codec_init(struct comp_dev *dev)
+{
+	struct codec_data *codec = comp_get_codec(dev);
+	struct waves_codec_data *waves_codec = NULL;
+	MaxxStatus_t status = 0;
+	int ret = 0;
+
+	comp_dbg(dev, "waves_codec_init() start");
+
+	waves_codec = codec_allocate_memory(dev, sizeof(struct waves_codec_data), 16);
+	if (!waves_codec) {
+		comp_err(dev, "waves_codec_init() allocate memory for waves_codec_data");
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	waves_codec_data_clear(waves_codec);
+	codec->private = waves_codec;
+
+	status = MaxxEffect_GetEffectSize(&waves_codec->effect_size);
+	if (status) {
+		comp_err(dev, "waves_codec_init() MaxxEffect_GetEffectSize() error %d",
+			status);
+		ret = -EIO;
+		codec_free_memory(dev, waves_codec);
+		goto err;
+	}
+
+	waves_codec->effect = (MaxxEffect_t *)codec_allocate_memory(dev,
+		waves_codec->effect_size, 16);
+
+	if (!waves_codec->effect) {
+		comp_err(dev, "waves_codec_init() allocate %d bytes for effect",
+			waves_codec->effect_size);
+		codec_free_memory(dev, waves_codec);
+		ret = -ENOMEM;
+		goto err;
+	}
+
+	comp_info(dev, "waves_codec_init() allocated %d bytes for effect",
+		waves_codec->effect_size);
+
+	comp_dbg(dev, "waves_codec_init() done");
+	return 0;
+
+err:
+	comp_err(dev, "waves_codec_init() failed %d", ret);
+	return ret;
+}
+
+int waves_codec_prepare(struct comp_dev *dev)
```

Comment:
should these be static ? i.e. are these ops for the codec adapter ?

Reply:
yes this should be static

Reply:
it is a part of codec_adapter interface, can`t be static

Reply:
@stolx we are discussing some related updates here at a special TSC - I will make sure you have an invite here since you are a user.

Reply:
ah yes, right i forgot the adapter declaration is not in this file, we will sync up with you